### ISSUE TITLE
Filetype detection by `vim.filetype.match` will fail in some cases when no buffer is specified

### ISSUE THREAD

**Issue by emonkak** [neovim/neovim#27265]
### Problem

I'm trying to detect the filetype using `vim.filetype.match()` based on the `filename`. However, I've noticed that it fails to detect some filetypes when no buffer is specified.

### Steps to reproduce

I expected a file with the ".ts" extension to be detected as TypeScript, but in fact, it might be detected as XML depending on the buffer content.

https://github.com/neovim/neovim/blob/4ffc20c9515294481486e81271a8edeeff203140/runtime/lua/vim/filetype.lua#L1031

In this case, when no `bur` is specified, the getting of buffer contents fails, and the filetype is not detected.

```lua
vim.filetype.match({ filename = 'index.ts', contents = {} }) -- => nil
```

https://github.com/neovim/neovim/blob/4ffc20c9515294481486e81271a8edeeff203140/runtime/lua/vim/filetype.lua#L2208-L2216

### Expected behavior

When no `buf` is specified, it is better to assume an empty buffer (or a buffer of `contents`) is passed and detect the filetype.

### Neovim version (nvim -v)

NVIM v0.10.0-dev-2244+g8bc578512

### Vim (not Nvim) behaves the same?

no

### Operating system/version

Linux 6.1.67-gentoo x86_64 GNU/Linux

### Terminal name/version

alacritty

### $TERM environment variable

xterm-256color

### Installation

build from repo

**Comment by clason** [comment]
This was intentional, although I forget the exact technical reason. If you think you can make it work (without failing all existing tests), PR welcome!

**Comment by clason** [comment]
Thinking more about it, I believe the situation was as follows:
* the dispatch mechanism (needed to support user-extensible content-aware checks) is built around actual buffers; making it work with buffers _or_ strings massively complicates a neuralgic API;
* this means one would have to open a scratch buffer with the provided contents, pass that to the dispatched detection function, and then clean up afterwards;
* this was deemed too much work for a use case that was considered niche and hence was left out until someone provided a compelling example of such a use case.

TL;DR: What's your use case here exactly?

**Comment by gpanders** [comment]
>the dispatch mechanism (needed to support user-extensible content-aware checks) is built around actual buffers; making it work with buffers or strings massively complicates a neuralgic API

This is correct, though I do think it might be possible to address cases like this where the buffer is invalid/does not exist and fallback to a default value.

I say "might be possible" because it will not always be obvious what the fallback/default value should be. In a case like this, with a `.ts` filename, "typescript" is the obvious choice for a default value over XML, but there are other filetypes where it's not as obvious. In those cases we need to decide what should be done.

**Comment by clason** [comment]
My preference would be "all or nothing" here: if there's a compelling use case for "contents but no buffer", then we should implement this properly (hopefully without affecting the general performance).

Otherwise we'd just be adding special cases on top of special cases...

**Comment by Julian** [comment]
Will just leave a note that I also found this issue now after being slightly surprised by the behavior when passing `contents =` just now in the context of writing a plugin which helps with finding test files across different languages given a source file (https://github.com/Julian/runt.nvim/), i.e. it implements "tell me the file path of the test file which corresponds to some file". In particular I know I do want an API that gets better results when a buffer is open containing the source file, but I also want an API that operates without that, for cases where a user directly wants to open a test file (without getting too far in the weeds on the workflow specifics). But in order to decide on language-specific behavior for where test files live, it was useful to use `vim.filetype.match` to get a filetype from the source file path, and the current behavior means I get to implement effectively what was described above about scratch buffers manually instead of e.g. reading in the contents of the source path and passing those along for the match.

(I'm not saying that in my opinion that means this justifies the complexity to support, just sharing my use case for taking into the equation).

**Comment by GitMurf** [comment]
FYI as seen in my linked issue above for a neovim plugin (https://github.com/olimorris/codecompanion.nvim/issues/722) I am having problems with vim.filetype.match() as well specifically for TypeScript (`*.ts`) files. `js` and `tsx` files work fine, but `*.ts` files always return `nil` even when trying to add the content with it as well.

See simple repro (note, I am on Windows):

```lua
lua vim.notify(vim.inspect(vim.filetype.match({ filename = 'testing.ts' })))
lua vim.notify(vim.inspect(vim.filetype.match({ filename = 'testing.js' })))
lua vim.notify(vim.inspect(vim.filetype.match({ filename = 'testing.tsx' })))
```
